db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/TriggerEventActivator.java

```diff
@@ -31,49 +31,78 @@
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Vector;
-
 import com.splicemachine.db.catalog.UUID;
 import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.jdbc.ConnectionContext;
+import com.splicemachine.db.iapi.services.context.ContextService;
 import com.splicemachine.db.iapi.services.io.FormatableBitSet;
 import com.splicemachine.db.iapi.sql.Activation;
+import com.splicemachine.db.iapi.sql.conn.ConnectionUtil;
 import com.splicemachine.db.iapi.sql.conn.LanguageConnectionContext;
 import com.splicemachine.db.iapi.sql.conn.StatementContext;
 import com.splicemachine.db.iapi.sql.dictionary.TriggerDescriptor;
 import com.splicemachine.db.iapi.sql.execute.CursorResultSet;
 
+import java.sql.SQLException;
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.splicemachine.db.impl.sql.execute.TriggerExecutionContext.pushLanguageConnectionContextToCM;
+
 /**
  * Responsible for firing a trigger or set of triggers based on an event.
  */
 public class TriggerEventActivator {
 
-    private LanguageConnectionContext lcc;
     private TriggerInfo triggerInfo;
     private TriggerExecutionContext tec;
     private Map<TriggerEvent, List<GenericTriggerExecutor>> statementExecutorsMap = new HashMap<>();
-    private Map<TriggerEvent, List<GenericTriggerExecutor>> rowExecutorsMap = new HashMap<>();
+    private Map<TriggerEvent, List<TriggerDescriptor>> rowExecutorsMap = new HashMap<>();
+    private Map<TriggerEvent, List<TriggerDescriptor>> rowConcurrentExecutorsMap = new HashMap<>();
     private Activation activation;
+    private ConnectionContext connectionContext;
     private String statementText;
     private UUID tableId;
     private String tableName;
-    private boolean tecPushed;
+    private boolean triggerExecutionContextPushed;
+    private boolean executionStmtValidatorPushed;
+
+    // getLcc() may return a different LanguageConnectionContext at the time
+    // we pop the triggerExecutionContext and executionStmtValidator, versus
+    // at the time they were pushed.  Saving the original "LCC" at the time of
+    // the push ensures that we pop from the correct LCC.  We sometimes need to
+    // save two versions LCCs because we may push to the LCC in the activation,
+    // and also to the LCC stored in the current ContextManager, if they happen
+    // to be different.  Trigger execution requires that both the activation and
+    // the current ContextManager both have a triggerExecutionContext.
+    private LanguageConnectionContext esvLCC1;
+    private LanguageConnectionContext esvLCC2;
+    private LanguageConnectionContext tecLCC1;
+    private LanguageConnectionContext tecLCC2;
```

Comment:
@OlegMazurov Do you think there's a chance of simplifying this now that we support nested context managers? Here's Mark's comment from the previous PR:

Yes it is messy. The ContextManager doesn't seem well designed for threading. The ExecutionStmtValidator is required to detect if we're doing DDL on a trigger table. The TriggerExecutionContext is required to be in both the LCC of the Activation and of the active ContextManager for proper trigger execution. A lot of the time the Activation and the ContextManager share the same LCC, which is what we want. But sometimes the ContextManager's LCC is pushed in someplace unexpected and we end up with a mismatch. I think the whole ContextManager logic should probably be revisited and maybe rewritten. I had tried some things to add more synchronization or use different data structures, e.g. ConcurrentHashMap in place of HashMap, but that either caused the database to run too slowly or caused hangs or other failures.

Reply:
I have filed _DB-9282: Rework DB-8883 to rely on nested ContextManager's_ to deal with this separately.